tools/circt-reduce/circt-reduce.cpp

```diff
@@ -11,13 +11,383 @@
+
+static cl::opt<std::string>
+    inputFilename(cl::Positional, cl::desc("<input file>"), cl::Required);
+
+static cl::opt<std::string>
+    outputFilename("o", cl::init("-"),
+                   cl::desc("Output filename for the reduced test case"));
+
+static cl::opt<bool>
+    keepBest("keep-best", cl::init(false),
+             cl::desc("Keep overwriting the output with better reductions"));
+
+static cl::opt<std::string> testerCommand(
+    "test", cl::Required,
+    cl::desc("A command or script to check if output is interesting"));
+
+static cl::list<std::string>
+    testerArgs("test-arg", cl::ZeroOrMore,
+               cl::desc("Additional arguments to the test"));
+
+//===----------------------------------------------------------------------===//
+// Tool Implementation
+//===----------------------------------------------------------------------===//
+
+/// An abstract reduction pattern.
+struct Reduction {
+  virtual ~Reduction() {}
+
+  /// Check if the reduction can apply to a specific operation.
+  virtual bool match(Operation *op) const = 0;
+
+  /// Apply the reduction to a specific operation. If the returned result
+  /// indicates that the application failed, the resulting module is treated the
+  /// same as if the tester marked it as uninteresting.
+  virtual LogicalResult rewrite(Operation *op) const = 0;
+
+  /// Return a human-readable name for this reduction pattern.
+  virtual std::string getName() const = 0;
+
+  /// Return true if the tool should accept the transformation this reduction
+  /// performs on the module even if the overall size of the output increases.
+  /// This can be handy for patterns that reduce the complexity of the IR at the
+  /// cost of some verbosity.
+  virtual bool acceptSizeIncrease() const { return false; }
+};
+
+/// A reduction pattern that applies an `mlir::Pass`.
+struct PassReduction : public Reduction {
+  PassReduction(MLIRContext *context, std::unique_ptr<Pass> pass,
+                bool canIncreaseSize = false)
+      : context(context), canIncreaseSize(canIncreaseSize) {
+    passName = pass->getArgument();
+    if (passName.empty())
+      passName = pass->getName();
+
+    if (auto opName = pass->getOpName())
+      pm = std::make_unique<PassManager>(context, *opName);
+    else
+      pm = std::make_unique<PassManager>(context);
+    pm->addPass(std::move(pass));
+  }
+  bool match(Operation *op) const override {
+    return op->getName().getIdentifier() == pm->getOpName(*context);
+  }
+  LogicalResult rewrite(Operation *op) const override { return pm->run(op); }
+  std::string getName() const override { return passName.str(); }
+  bool acceptSizeIncrease() const override { return canIncreaseSize; }
+
+protected:
+  MLIRContext *const context;
+  std::unique_ptr<PassManager> pm;
+  StringRef passName;
+  bool canIncreaseSize;
+};
+
+/// A sample reduction pattern that maps `firrtl.module` to `firrtl.extmodule`.
+struct ModuleExternalizer : public Reduction {
```

Comment:
I'd recommend splitting this stuff out to a .h (containing the `Reduction` base class) and .cpp file in tools/circt-reduce to better structure the code.

Reply:
That totally makes sense. Especially since we'd later on probably want to have dialects themselves provide these `Reduction` classes, so they'll want to have a proper interface in a header file somewhere.